### PR TITLE
support dynamically registering/deregistering as a consul service

### LINKED ISSUE #1988: sighup doesnt reload consul.register_service
**Describe the bug**
The register_service setting for Consul isn't dynamically adjustable.

**To Reproduce**
1. Start Patroni with Consul as the DCS and without setting consul.register_service in patroni.yml, so it defaults to false
2. set consul.register_service: true in patroni.yml
3. reload patroni (either by sighup or by the reload endpoint)

**Expected behavior**
I'd expect this to enable Consul service registration

**Environment**
- Patroni version: 2.0.2
- PostgreSQL version: 13..3
- DCS (and its version): Consul 1.10.0

I'm happy to put together a PR if this is something that yall are interested in accepting

### PR DESCRIPTION
**Pull request by tommyzli**
closes #1988

### CODE REVIEW

**Review by CyberDem0n — COMMENTED** [review]
Changes are looking good to me, but I have a little question.
Shouldn't we also handle the change of `service_tags` dynamically?

**Comment by tommyzli** [comment]
Ah right, will add

**Comment by tommyzli** [comment]
hey @CyberDem0n, mind taking a second look?

**Review by CyberDem0n — COMMENTED** [review]
Tags reload doesn't work:
1. I started Patroni with `service_tags: [foo, buzz]`
```
2021-07-14 15:13:28.721 CEST [23459] LOG:  database system is ready to accept connections
2021-07-14 15:13:29,677 INFO: Lock owner: postgresql0; I am postgresql0
2021-07-14 15:13:29,711 INFO: Register service batman, params {'check': {'DeregisterCriticalServiceAfter': '150.0s', 'interval': '5s', 'http': 'http://127.0.0.1:8008/master'}, 'service_id': 'batman/postgresql0', 'tags': ['bar', 'foo', 'master'], 'port': 5432, 'address': '127.0.0.1'}
2021-07-14 15:13:29,739 INFO: no action. I am (postgresql0) the leader with the lock
2021-07-14 15:13:29,767 INFO: no action. I am (postgresql0) the leader with the lock
```
2. changed `service_tags` to `[foo, buzz]`, sent a SIGHUP and nothing happened:
```
2021-07-14 15:14:49,757 INFO: no action. I am (postgresql0) the leader with the lock
2021-07-14 15:14:59,757 INFO: no action. I am (postgresql0) the leader with the lock
2021-07-14 15:15:09,760 INFO: Reloading PostgreSQL configuration.
server signaled
2021-07-14 15:15:09.764 CEST [23459] LOG:  received SIGHUP, reloading configuration files
2021-07-14 15:15:10,769 INFO: no action. I am (postgresql0) the leader with the lock
2021-07-14 15:15:19,756 INFO: no action. I am (postgresql0) the leader with the lock
```

Consul API also reports old tags:
```json
curl -s localhost:8500/v1/agent/services | jq .
{                                                                                                                                                                                                                                                                               
  "batman/postgresql0": {                                                                                                                                                                                                                                                       
    "ID": "batman/postgresql0",                                                                                                                                                                                                                                                 
    "Service": "batman",
    "Tags": [
      "bar",
      "foo",
      "master"
    ],
    "Meta": {},
    "Port": 5432,
    "Address": "127.0.0.1",
    "Weights": {
      "Passing": 1,
      "Warning": 1
    },
    "EnableTagOverride": false
  }
}

```

**Comment by tommyzli** [comment]
Apparently by default, Consul silently ignores changes to service tags. Adding `enable_tag_override: True` to the service definition resolves that

File: patroni/dcs/consul.py [hunk 1]
```diff
@@ -229,7 +229,8 @@ def __init__(self, config):
         self._register_service = config.get('register_service', False)
         self._previous_loop_register_service = self._register_service
         if self._register_service:
-            self._service_tags = config.get('service_tags', [])
+            self._service_tags = sorted(config.get('service_tags', []))
+            self._previous_loop_service_tags = self._service_tags
```
Comment (CyberDem0n) [inline patroni/dcs/consul.py:233]: If upon start the `register_service` isn't set in the config, the next call of `reload_config()` will fail due to undefined `self._service_tags`.

**Review by CyberDem0n — COMMENTED** [review]
> Adding enable_tag_override: True to the service definition resolves that

No, it doesn't. The issue is that the `agent.service.register()` method isn't called. Even the `Consulupdate_service()` isn't called, probably due to the https://github.com/zalando/patroni/pull/1993/files#diff-99372e752231ac1fddf9cda076dd985f0ed8293e8ba1c06ea1d95a0530ccb959R421

**Comment by tommyzli** [comment]
Oh, I see what happened. I've mostly been testing this on short lived patroni processes, and during the first minutes of its life the [Member](https://github.com/zalando/patroni/blob/master/patroni/dcs/consul.py#L394) data changes a few times as the node gets promoted from standby to leader
That causes this [short circuit](https://github.com/zalando/patroni/blob/master/patroni/dcs/consul.py#L405) to do nothing because `deep_compare` indicates a change, and as a side effect allows the config reload to correctly change the service definition since it happened to get wrapped up in the same loop iteration

Question for you @CyberDem0n - how important do you think that short circuiting logic is for Consul? I checked the other DCS implementations and it's not consistent. Raft and Etcd will write the key each time, and the others will exit early if no data has changed.

I can remove the short circuit. Or I can work around the logic, it just might result in some hard to read code. Up to you

**Comment by CyberDem0n** [comment]
> how important do you think that short circuiting logic is for Consul? I checked the other DCS implementations and it's not consistent. Raft and Etcd will write the key each time, and the others will exit early if no data has changed.

It all depends on how the expiration mechanism is implemented. Etcdv3, Consul, Zookeeper - we are using session mechanisms, and the member key is associated with a sessionId. If the session expires it causes also associated keys to expire. But, the sessions mechanism also allows us not to update keys if the value didn't change. In case of Etcdv2 or Raft every key has its own TTL, and therefore they have to be updated all the time.

IMO, it is better to avoid doing unnecessary writes to DCS, therefore short-circuting was implemented.

**Comment by tommyzli** [comment]
A bit of duplicated logic, but now it does the correct thing:
```
# upon startup, with service_tags = ['baz', 'biz']
2021-07-21 14:34:38,648 INFO: Register service batman, params {'service_id': 'batman/postgresql0', 'address': '127.0.0.1', 'port': 5432, 'check': {'http': 'http://127.0.0.1:8008/master', 'interval': '5s', 'DeregisterCriticalServiceAfter': '150.0s'}, 'tags': ['baz', 'biz', 'master'], 'enable_tag_override': True}
...

# removed service_tags
2021-07-21 14:36:08.717 PDT [41668] LOG:  received SIGHUP, reloading configuration files
2021-07-21 14:36:09,726 INFO: Lock owner: postgresql0; I am postgresql0
2021-07-21 14:36:09,728 INFO: Register service batman, params {'service_id': 'batman/postgresql0', 'address': '127.0.0.1', 'port': 5432, 'check': {'http': 'http://127.0.0.1:8008/master', 'interval': '5s', 'DeregisterCriticalServiceAfter': '150.0s'}, 'tags': ['master'], 'enable_tag_override': True}
2021-07-21 14:36:09,770 INFO: no action. I am (postgresql0) the leader with the lock
2021-07-21 14:36:18,699 INFO: Lock owner: postgresql0; I am postgresql0
...

# turned off register_service
2021-07-21 14:46:26.018 PDT [41989] LOG:  received SIGHUP, reloading configuration files
2021-07-21 14:46:27,027 INFO: Lock owner: postgresql0; I am postgresql0
2021-07-21 14:46:27,030 INFO: Deregister service batman/postgresql0
2021-07-21 14:46:27,033 INFO: no action. I am (postgresql0) the leader with the lock
```

Other odd combinations look good as well (starting with service_tags defined but not register_service does not register, turning on register_service afterwards registers with previously set tags, etc) 
I feel like this has become complex enough to warrant it's own suite of behave tests 😅

File: patroni/dcs/consul.py [hunk 2]
```diff
@@ -402,14 +412,23 @@ def touch_member(self, data, permanent=False):
...
+        if self._register_service or self._previous_loop_register_service:
+            try:
+                self.update_service(not create_member and member and member.data or {}, data)
+            except InvalidSession:
+                self._session = None
+                logger.error('Our session disappeared from Consul, can not "touch_member"')
+                return False
+            except Exception:
+                logger.exception('touch_member')
+                return False
```
Comment (CyberDem0n) [inline patroni/dcs/consul.py:424]: The `update_service()` isn't supposed to raise exceptions, at least not the `InvalidSession`.

```suggestion
            try:
                self.update_service(not create_member and member and member.data or {}, data)
            except Exception:
                logger.exception('update_service')
```

**Review by CyberDem0n — COMMENTED** [review]
> I feel like this has become complex enough to warrant it's own suite of behave tests sweat_smile

Well, yeah... But so far behave tests are not really DCS specific.

**Comment by FxKu** [comment]
:+1:

**Comment by CyberDem0n** [comment]
:+1:

**Comment by CyberDem0n** [comment]
Merged, thank you @tommyzli!

**Pull request merged**